miner/src/service_transaction_checker.rs

```diff
@@ -16,37 +16,111 @@
 
 //! A service transactions contract checker.
 
-use call_contract::{CallContract, RegistryInfo};
+use std::collections::BTreeMap;
+use call_contract::{RegistryInfo, CallContract};
+use chain_info::ChainInfo;
 use types::ids::BlockId;
 use types::transaction::SignedTransaction;
 use ethabi::FunctionOutputDecoder;
+use ethereum_types::Address;
+use parking_lot::RwLock;
 
 use_contract!(service_transaction, "res/contracts/service_transaction.json");
 
 const SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME: &'static str = "service_transaction_checker";
 
+/// Certified addresses cache
+#[derive(Default)]
+struct CertifiedAddressesCache {
+	cache: RwLock<BTreeMap<Address, bool>>,
+	last_block: u64
+}
+
 /// Service transactions checker.
 #[derive(Default, Clone)]
-pub struct ServiceTransactionChecker;
+pub struct ServiceTransactionChecker {
+	certified_addresses_cache: CertifiedAddressesCache
+}
 
 impl ServiceTransactionChecker {
-	/// Checks if given address is whitelisted to send service transactions.
-	pub fn check<C: CallContract + RegistryInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
+	/// Checks if given address in tx is whitelisted to send service transactions.
+	pub fn check<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
 		let sender = tx.sender();
-		let hash = tx.hash();
-
 		// Skip checking the contract if the transaction does not have zero gas price
 		if !tx.gas_price.is_zero() {
 			return Ok(false)
 		}
 
-		let address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
+		self.check_address(client, sender)
+	}
+
+	/// Checks if given address is whitelisted to send service transactions.
+	pub fn check_address<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, sender: Address) -> Result<bool, String> {
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if (client.chain_info().best_block_number as u64 == self.certified_addresses_cache.last_block) && cache.is_some() {
```

Comment:
to avoid `unwrap` in the second line you can try this:
```
if client.chain_info().best_block_number == self.cache_last_block {
   let cache = self.cache.cache.try_read();
   if let Some(allowed) = cache.and_then(|x| x.get(&sender)) {
        return Ok(*allowed);
   }
}
```

Reply:
Actually calling `chain_info()` might be quite heavy, let's just ditch that?

Reply:
I agree, once it's possible to guarantee that the cache is 100% updated on the new block we can get rid of `cache_last_block `.

Reply:
I'd go for `try_read()` here and `.write()` when cache is updated - that way instead of blocking for the cache to be ready we will just do additional state call, which shouldn't really be that heavy.